core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -340,20 +343,54 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
         final RoutingResult routingResult = routed.routingResult();
         final ServiceConfig serviceCfg = routed.value();
         final HttpService service = serviceCfg.service();
-
+        final EventLoop serviceEventLoop;
+        final EventLoopGroup serviceWorkerGroup = serviceCfg.serviceWorkerGroup();
+        if (serviceWorkerGroup == config.workerGroup()) {
+            serviceEventLoop = eventLoop;
+        } else {
+            serviceEventLoop = serviceWorkerGroup.next();
+        }
         final DefaultServiceRequestContext reqCtx = new DefaultServiceRequestContext(
-                serviceCfg, channel, config.meterRegistry(), protocol,
+                serviceCfg, channel, serviceEventLoop, config.meterRegistry(), protocol,
                 nextRequestId(), routingCtx, routingResult, req.exchangeType(),
                 req, sslSession, proxiedAddresses, clientAddress,
                 System.nanoTime(), SystemInfo.currentTimeMicros());
 
+        final boolean needsDirectExecution = serviceEventLoop == config.workerGroup();
+
+        if (needsDirectExecution) {
+            serveRequest(ctx, req, serviceCfg, service, routingCtx, routingResult,
+                         routed, eventLoop, reqCtx, serviceWorkerGroup);
+        } else {
+            serviceEventLoop.execute(() -> serveRequest(ctx, req, serviceCfg, service, routingCtx,
```

Comment:
Question) Maybe I'm missing something, but why do we schedule on `serviceWorkerGroup` twice?

1) https://github.com/line/armeria/blob/1e180096f0b7a2e6f7af5c424a7d3bfb0446bba3/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L365-L367

2) https://github.com/line/armeria/blob/1e180096f0b7a2e6f7af5c424a7d3bfb0446bba3/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L387-L393

From my understanding, we can't guarantee that `serviceEventLoop` is always run from the same thread (since `EventLoopGroup` maintains multiple threads).
Hence, I think it's also possible that the following access to `HashMap` for instance now should consider multiple threads.

https://github.com/line/armeria/blob/1e180096f0b7a2e6f7af5c424a7d3bfb0446bba3/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L419

@ other-maintainers, what do you think of isolating the range executed by `serviceEventLoop` to `service.serve(ctx, req)`? We can keep concerns of multithreaded access to a minimum.

```
private void handleRequest(...) {
...
  final DefaultServiceRequestContext reqCtx =...
  serveRequest(ctx, req, serviceCfg, service, routingCtx, routingResult, routed, eventLoop, reqCtx, 
  serviceEventLoop);
}
...

private void serveRequest(ChannelHandlerContext ctx,
...
  serviceResponse = HttpResponse.from(() -> {
    try {
      return service.serve(reqCtx, req);
    } catch (Exception cause) {
...
    }
  }, serviceEventLoop);
...
}
```

Reply:
I suggested using number 1 to avoid wrapping the returned `HttpResponse` with  `DeferredHttpResponse`. So we should remove `HttpResponse.from(() -> {  ... }, serviceWorkerGroup)`.

Reply:
To conclude, we have two options:
- Isolating the range executed by `serviceEventLoop` to `service.serve(ctx, req)`. [(sample code)](https://github.com/line/armeria/pull/4136#discussion_r941248280)
- Delegating all execution to the `serviceEventLoop` from `service.serve(ctx, req)`.
  ```java
  if (needsDirectExecution) {
      serveRequest(...);
  else {
      serviceEventLoop.execute(() -> serveReqeust(...));
  }

  private void serveRequest(...) {
      try (SafeCloseable ignored = reqCtx.push()) {
          ...
          HttpResponse serviceResponse;
          try {
              req.init(reqCtx);
              serviceResponse = service.serve(reqCtx, req);
          catch (...) {
              ...
          }
          ...
      }
  }
  ```

Pros and cons of the first option:
- We can minimize the code change.
- We need to use `HttpResponse.from(...)` which has an overhead.

Pros and cons of the second option:
- We don't have to use `HttpResponse.from(...)`. The `serviceEventLoop` handles the rest of the logic.
- We need to make the code thread safe because multiple threads can now access the code. (Which also brings another overhead)
  - `unfinishedRequests` 
  - The implementations of `DecodedHttpRequest.setResponse(res)`.
  
Because this is a critical path of the server's requests, I think minimizing overhead is the most crucial factor.
So we need a benchmark result for each option.
If the result is not so much different, I personally would like to go with the first option.

Reply:
If it's too much for the PR, please let us know. I don't want to give you a burden. 😄
You can just implement whichever option you want and we can do the benchmark later.

Reply:
I would like to go with the simpler option for this iteration so that we can focus on correctness and minimizing bugs for now.

Although I agree this is a critical path, 
1. For most cases `serviceWorkerGroup == workerGroup`
2. `serviceWorkerGroup` is an opt-in feature

So I propose that we go with the simpler option for now.